added most of simple_counting_scope (#36)
ispeters/async_scope: asyncscope.md

    assoc(assoc&& other) noexcept;
    ~assoc();
    assoc& operator=(assoc rhs) noexcept;

    explicit operator bool() const noexcept;

  private:
    simple_counting_scope* @_scope_@{}; // @_exposition-only_@
  };

  struct @_join-t_@; // @_exposition-only_@

  enum @_state-type_@ { // @_exposition-only_@
    @_unused_@, // @_exposition-only_@
    @_open_@, // @_exposition-only_@
    @_close_@, // @_exposition-only_@
    @_open-and-joining_@, // @_exposition-only_@
    @_closed-and-joining_@, // @_exposition-only_@
    @_unused-and-closed_@, // @_exposition-only_@
    @_joined_@, // @_exposition-only_@
  };

  // [exec.scounting.ctor], constructor and destructor
  simple_counting_scope() noexcept;
  simple_counting_scope(simple_counting_scope&&) = delete;
  ~simple_counting_scope();

  // [exec.scounting.mem], members
  token get_token() noexcept;
  void close() noexcept;
  auto join() noexcept;

private:
  size_t @_count_@; // @_exposition-only_@
  @_state-type_@ @_state_@; // @_exposition-only_@
};
```

[1]{.pnum} A `simple_counting_scope` maintains a count of outstanding
    operations.  Let `s` be an object of type `simple_counting_scope`,

>     `t` be an object of type `token` obtained from `s.get_token()`,

Reviewer: Does `token` need to be qualified here?
```suggestion
    `t` be an object of type `simple_counting_scope::token` obtained from `s.get_token()`,
```

Author: I _think_ we normally don't qualify these although in an actual implementation we would. I don't mind either way. I can add the qualification and take it out if LWG complains.

    started while the `s` is in `@_unused_@` or `@_open_@` state, `s`
    moves to the `@_open-and-joining_@` state.
- [1.4]{.pnum} `@_closed_@`: when `s.close()` is called while `s` is
    in `open` state, `s` moves to the `@_closed_@` state.
- [1.5]{.pnum} `@_unused-and-closed_@`: when `s.close()` is called while
    `s` is in `@_unused_@` state, `s` moves to the `@_unused-and-closed_@`
    state.
- [1.6]{.pnum} `@_closed-and-joining_@`: when `s.close()` is called while
    `s` is in `@_open-and-joining_@` state or the operation state `o` is 
    started while `s` is in `@_closed_@` or `@_unused-and-close_@` state, `s`
    moves to the `@_closed-and-joining_@` state.
- [1.7]{.pnum} `@_joined_@`: when the count of associated objects drops to
    zero while `s` is in `@_open-and-joining_@` or `@_closed-and-joining_@`
    state, `s` moves to the `@_joined_@` state.

##### Constructor and Destructor [exec.scounting.ctor]

`simple_counting_scope() noexcept;`

[1]{.pnum} _Postcondtions:_ `@_count_@` is `0` and `@_state_@` is `@_unused_@`

`~simple_counting_scope();`

[2]{.pnum} _Effects:_ If `@_state_@` is not one of `@_joined_@`, `@_unused_@`,

>     or `@_unused-and-close_@`, invokes `terminate` ([except.terminate]{.sref}).

Reviewer: ```suggestion
    or `@_unused-and-closed_@`, invokes `terminate` ([except.terminate]{.sref}).
```

Author: I fixed that instance, the other your noticed, and I think I found two more...

    Otherwise, has no effects.

##### Members [exec.scounting.mem]

`token get_token() noexcept;`

> [1]{.pnum} _Returns:_ An object `t` of type token such that

Reviewer: Should "token" be in code font? And should it be qualified?
```suggestion
[1]{.pnum} _Returns:_ An object `t` of type `simple_counting_scope::token` such that
```

    `t.@_scope_@ == this` is `true`.

`void close() noexcept;`

[2]{.pnum} _Efects:_ If `@_state_@` is

> - [2.1]{.pnum} `@_unused_@` changes `@_state_@` to `@_unused-and_close_@`;

Reviewer: ```suggestion
- [2.1]{.pnum} `@_unused_@` changes `@_state_@` to `@_unused-and_closed_@`;
```

Author: right. ... and I should also change "and_closed" to "and-closed".

- [2.2]{.pnum} `@_open_@` changes `@_state_@` to `@_closed_@`;
- [2.3]{.pnum} `@_open-and-joining_@` changes `@_state_@` to `@_closed-and-joining_@`;

> [3]{.pnum} Any call to `t.try_associate()` for a `token` object `t`
>     referring to a `simple_counting_scope` object `s` which are
>     sequenced after a call to `s.close()` return a disengaged
>     association ([exec.scounting.assoc]{.sref}, p1).

Reviewer: ```suggestion
[3]{.pnum} Any call to `t.try_associate()` for a `token` object `t`
    referring to a `simple_counting_scope` object `s` which is
    sequenced after a call to `s.close()` returns a disengaged
    association ([exec.scounting.assoc]{.sref}, p1).
```

`auto join() noexcept;`

[4]{.pnum} _Returns:_ `@_make_sender_@(@_join-t_@, this)`

[5]{.pnum} The exposition-only class template `@_impls-for_@`
    ([exec.snd.general]{.sref}) is specialized for `@_join-t_@`
    as follows:

```

>     template <>
>     struct @_impls-for_@<@_join-t_@>: @_default-impls_@> {
>         template <class Receiver>
>         struct @_state_@ {  // @_exposition-only_@
>             simple_counting_scope* @_scope_@; // @_exposition-only_@
>             remove_cvref_t<Receiver>& @_receiver_@; // @_exposition-only_@
> 
>             void @_complete_@() { // @_exposition-only_@
>                 set_value(std::move(@_receiver_@));
>             }
>         };
> 
>         static constexpr auto @_get-state_@ =
>           []<class Receiver>(auto&& sender, Receiver& receiver) {
>             auto[_, self] = sender;
>             return @_state_@<Receiver>(self, receiver);
>           };
>         static constexpr auto @_start_@ =
>             [](auto& s, auto&) { @_see-below_@; };
>     };
> ```
> 
> [7]{.pnum} In the function object used to initialize
>     `@_impls-for_@<@_join-t_@>::@_start_@` let `state` be `s.@_scope_@->@_state_@`.
>     If `state` is
> 
> - [7.1]{.pnum} `@_unused_@`, `@_unused-and-closed_@`, or `@_joined_@`,
>     `s.@_complete_@()` is invoked;
> - [7.2]{.pnum} `@_open_@`, changes the state of `*s.@_scope_@` to `@_open-and-joining_@`;
> - [7.3]{.pnum} `@_closed_@`, changes the state of `*s.@_scope_@` to `@_closed-and-joining_@`;
> 
> [8]{.pnum} If `s.@_complete_@()` was not invoked, registers `s` with
>     `*s.@_scope_@` to have `s.@_complete_@()` invoked when `s.@_scope_@->@_count_@` becomes
>     zero.

Reviewer: I'm not sure I understand how _`impls-for`_ works, but I don't think this is sufficient; the join-sender should complete on the execution context associated with `get_scheduler(receiver)` if it doesn't complete inline in `start()`.

Also, I think that starting the join-sender in the _`unused`_ or _`unused-and-closed`_ state should probably move the state to _`joined`_ so that no future `try_associate()` calls can succeed.

```suggestion
    template <>
    struct @_impls-for_@<@_join-t_@>: @_default-impls_@> {
        template <class Receiver>
        struct @_state_@ {  // @_exposition-only_@
            simple_counting_scope* @_scope_@; // @_exposition-only_@
            remove_cvref_t<Receiver>& @_receiver_@; // @_exposition-only_@
            using @_op_t_@ = decltype(connect(schedule(get_scheduler(receiver)), receiver)); // @_exposition-only_@
            op_t @_op_@; // @_exposition-only_@

            @_state_@(simple_counting_scope* scope, Receiver& receiver) // @_exposition-only_@
              : @_scope_@(scope),
                @_receiver_@(receiver),
                @_op_@(connect(schedule(get_scheduler(receiver)), receiver)) {}

            void @_complete_@() { // @_exposition-only_@
                @_op_@.start();
            }

            void @_complete-inline_@() { // @_exposition-only_@
                set_value(std::move(@_receiver_@));
            }
        };

        static constexpr auto @_get-state_@ =
          []<class Receiver>(auto&& sender, Receiver& receiver) {
            auto[_, self] = sender;
            return @_state_@<Receiver>(self, receiver);
          };
        static constexpr auto @_start_@ =
            [](auto& s, auto&) { @_see-below_@; };
    };
```

[7]{.pnum} In the function object used to initialize
    `@_impls-for_@<@_join-t_@>::@_start_@` let `state` be `s.@_scope_@->@_state_@`.
    If `state` is

- [7.1]{.pnum} `@_unused_@`, `@_unused-and-closed_@`, or `@_joined_@`,
    `s.@_complete-inline_@()` is invoked and changes the state of `*s.@_scope_@` to `@_joined_@`;
- [7.2]{.pnum} `@_open_@`, changes the state of `*s.@_scope_@` to `@_open-and-joining_@`;
- [7.3]{.pnum} `@_closed_@`, changes the state of `*s.@_scope_@` to `@_closed-and-joining_@`;

[8]{.pnum} If `s.@_complete-inline_@()` was not invoked, registers `s` with
    `*s.@_scope_@` to have `s.@_complete_@()` invoked when `s.@_scope_@->@_count_@` becomes
    zero.
```

Author: Yes, I didn't have on the radar that the join-sender needs to take the result of `get_scheduler(get_env(receiver))` into account. I think that change makes sense. When I looked at it I first thought that returning `continues_on(some-sender)` would be the thing but I like your approach better.

Note that there is a `get_env(receiver)` needed: the receivers are not required to respond to queries. I _think_ receiver can be their own environments but generic code need to travel via the environment.

##### Token [exec.scounting.token]

`template <sender Sender>`
`Sender&& wrap(Sender&& snd) const noexcept`;

[1]{.pnum} _Returns:_ `std::forward<Sender>(snd);`

`assoc try_associate() const;`

[2]{.pnum} _Effects:_ **TODO**

[3]{.pnum} _Returns:_ **TODO**

##### Assoc [exec.scounting.assoc]

[1]{.pnum} An object `a` of type `assoc` is _disengaged_ if `a.@_scope_@
    == nullptr` is `true` and _engaged_ otherwise.

`assoc(const assoc& other) noexcept;`

[2]{.pnum} _Effects_: If `other` is engaged evalutes
    `*this = other.@_scope_@->get_token().try_associate()`.

`assoc(assoc&& other) noexcept;`

[3]{.pnum} _Effects_: Initializes `@_scope_@` with
    `exchange(other.@_scope_@, nullptr)`.

`~assoc();`

[4]{.pnum} _Effects_: If the object is disengaged, does nothing;

>     otherwise decrements `@_scope_@->@_count_@`. If `@_scope_@->@_count_@`
>     is zero after decrementing, changes the state of `*@_scope_@`
>     to `@_joined_@` and calls `@_complete_@()` on all objects
>     registered with `*@_scope_@`.

Reviewer: We only transition to _`joined`_ from _`open-and-joining`_ and _`closed-and-joining`_.

```suggestion
    otherwise decrements `@_scope_@->@_count_@`. If `@_scope_@->@_count_@`
    is zero after decrementing and `@_scope_@->@_state_@` is `@_open-and-joining_@`
    or `@_closed-and-joining_@`, changes the state of `*@_scope_@` to `@_joined_@`
    and calls `@_complete_@()` on all objects registered with `*@_scope_@`.
```

Author: OK. That should be fine. I should also have a look at the state diagram.

## `execution::simple_counting_scope`

Add the following new section imediate after **TODO**:

::: add
### Scopes [exec.scopes]

#### Simple Counting Scope [exec.simple.counting.scope]

##### General [exec.scounting.general]

```

> class simple_counting_scope {

Reviewer: What's the Committee's policy on using `struct` or `class`? In section 5, this is a `struct`, but that's basically just my whim; if `class` is expected, we should probably update section 5 to match.

Author: I think entities which are bags of data like `pair` or customisation objects would be `struct`s. Things with `private` members would be `class`. That's easily changed anyway.

[1]{.pnum} _Returns:_ An object `t` of type `simple_counting_scope::token` such that
    `t.@_scope_@ == this` is `true`.

`void close() noexcept;`

[2]{.pnum} _Efects:_ If `@_state_@` is

- [2.1]{.pnum} `@_unused_@` changes `@_state_@` to `@_unused-and-closed_@`;
- [2.2]{.pnum} `@_open_@` changes `@_state_@` to `@_closed_@`;
- [2.3]{.pnum} `@_open-and-joining_@` changes `@_state_@` to `@_closed-and-joining_@`;

[3]{.pnum} Any call to `t.try_associate()` for a `token` object `t`
    referring to a `simple_counting_scope` object `s` which is
    sequenced after a call to `s.close()` returns a disengaged
    association ([exec.scounting.assoc]{.sref}, p1).

`auto join() noexcept;`

[4]{.pnum} _Returns:_ `@_make_sender_@(@_join-t_@, this)`

[5]{.pnum} The exposition-only class template `@_impls-for_@`
    ([exec.snd.general]{.sref}) is specialized for `@_join-t_@`
    as follows:

```
template <>
struct @_impls-for_@<@_join-t_@>: @_default-impls_@> {
    template <class Receiver>
    struct @_state_@ {  // @_exposition-only_@
        simple_counting_scope* @_scope_@; // @_exposition-only_@
        remove_cvref_t<Receiver>& @_receiver_@; // @_exposition-only_@

>         using @_op_t_@ = decltype(connect(schedule(get_scheduler(receiver)), receiver)); // @_exposition-only_@

Reviewer: ```suggestion
        using @_op_t_@ = decltype(connect(schedule(get_scheduler(get_env(receiver))), receiver)); // @_exposition-only_@
```

        @_op_t_@ @_op_@; // @_exposition-only_@

        @_state_@(simple_counting_scope* scope, Receiver& receiver) // @_exposition-only_@
          : @_scope_@(scope),
            @_receiver_@(receiver),

>             @_op_@(connect(schedule(get_scheduler(receiver)), receiver)) {}

Reviewer: ```suggestion
            @_op_@(connect(schedule(get_scheduler(get_env(receiver))), receiver)) {}
```

    Sender&& wrap(Sender&& snd) const noexcept;
    assoc try_associate() const;

  private:
    simple_counting_scope* @_scope_@; // @_exposition-only_@
  };

  // [exec.scounting.assoc], assoc
  struct assoc {
    assoc() noexcept = default;
    assoc(const assoc& other) noexcept;
    assoc(assoc&& other) noexcept;
    ~assoc();
    assoc& operator=(assoc rhs) noexcept;

    explicit operator bool() const noexcept;

  private:
    simple_counting_scope* @_scope_@{}; // @_exposition-only_@
  };

  struct @_join-t_@; // @_exposition-only_@

  enum @_state-type_@ { // @_exposition-only_@
    @_unused_@, // @_exposition-only_@
    @_open_@, // @_exposition-only_@
    @_close_@, // @_exposition-only_@
    @_open-and-joining_@, // @_exposition-only_@
    @_closed-and-joining_@, // @_exposition-only_@
    @_unused-and-closed_@, // @_exposition-only_@
    @_joined_@, // @_exposition-only_@
  };

  // [exec.scounting.ctor], constructor and destructor
  simple_counting_scope() noexcept;
  simple_counting_scope(simple_counting_scope&&) = delete;
  ~simple_counting_scope();

  // [exec.scounting.mem], members
  token get_token() noexcept;
  void close() noexcept;

>   auto join() noexcept;

Reviewer: Is it worth constraining this `auto`?
```suggestion
  sender auto join() noexcept;
```

Author: I made it consistent with the class definition. For `counting_scope` the member `join()` is specified to return a `join-sender`: maybe that should be done here as well?

[3]{.pnum} _Returns:_ An engaged `assoc` object `a` with
    `a.@_scope_@ == @_scope_@` if `@_scope_@->@_count_@` was
    incremented, a disengaged `assoc` object otherwise.

##### Assoc [exec.scounting.assoc]

[1]{.pnum} An object `a` of type `assoc` is _disengaged_ if
    `a.@_scope_@ == nullptr` is `true` and _engaged_ otherwise.

`assoc(const assoc& other) noexcept;`

[2]{.pnum} _Effects_: If `other` is engaged evaluates
    `*this = other.@_scope_@->get_token().try_associate()`.

`assoc(assoc&& other) noexcept;`

[3]{.pnum} _Effects_: Initializes `@_scope_@` with
    `exchange(other.@_scope_@, nullptr)`.

`~assoc();`

[4]{.pnum} _Effects_: If the object is disengaged, does nothing;
    otherwise decrements `@_scope_@->@_count_@`. If `@_scope_@->@_count_@`
    is zero after decrementing and `@_scope_@->@_state_@` is `@_open-and-joining_@`
    or `@_closed-and-joining_@`, changes the state of `*@_scope_@` to `@_joined_@`
    and calls `@_complete_@()` on all objects registered with `*@_scope_@`.

[5]{.pnum} [_Note:_ Calling `@_complete_@()` on any registered object
    may cause `*@_scope_@` to get destroyed.  _--End-Note_]

`assoc& operator=(assoc rhs) noexcept;`

> [6]{.pnum} _Effects_: Equivalent to `swap(@_scope_@, other.@_rhs_@)`.

Reviewer: ```suggestion
[6]{.pnum} _Effects_: Equivalent to `swap(@_scope_@, rhs.@_scope_@)`.
```

Author: done
